rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+    `/foo/bar/flake` is equivalent to `/foo/bar?dir=flake`.
+
+* GitHub repositories. These are downloaded as tarball archives,
+  rather than through Git. This is often much faster and uses less
+  disk space since it doesn't require fetching the entire history of
+  the repository. On the other hand, it doesn't allow incremental
+  fetching (but full downloads are often faster than incremental
+  fetches!). The syntax is:
+
+      github:<owner>/<repo>(/<rev-or-ref>)?(\?<params>)?
+
+  `<rev-or-ref>` specifies the name of a branch or tag (`ref`), or a
+  commit hash (`rev`). Note that unlike GitHub, Git allows fetching by
+  commit hash without specifying a branch or tag.
+
+  The only supported parameter is `dir` (see above).
+
+  Some examples:
+
+  * `github:edolstra/dwarffs`
+  * `github:edolstra/dwarffs/unstable`
+  * `github:edolstra/dwarffs/d3f2baba8f425779026c6ec04021b2e927f61e31`
+
+* Indirections through the flake registry. These have the form
+
+      <flake-id>(/<rev-or-ref>(/rev)?)?
+
+  These perform a lookup of `<flake-id>` in the flake registry. The
+  specified `rev` and/or `ref` are then merged with the entry in the
+  registry. (See below.) For example, `nixpkgs` and
+  `nixpkgs/release-19.03` are indirect flake references.
+
+In the future, we should also add tarball flake references
+(e.g. `https://example.org/nixpkgs.tar.xz`). It would also be
+straight-forward to add flake references to Mercurial repositories
+since Nix already has Mercurial support.
+
+## Flake registries
+
+Flake registries map symbolic flake identifiers (e.g. `nixpkgs`) to
+"direct" flake references (i.e. any type of flake reference that's not
+an indirection). This is a convenience to users, allowing them to do
+
+    nix run nixpkgs:hello
+
+rather than
+
+    nix run github:NixOS/nixpkgs:hello
+
+There are multiple registries:
+
+* The global registry
+  https://raw.githubusercontent.com/NixOS/flake-registry/master/flake-registry.json. (This
+  location can be overriden via the `flake-registry` option.) Nix
+  automatically fetches this registry periodically. The check interval
+  is determined by the `tarball-ttl` option.
+
+* The local registry `~/.config/nix/registry.json`.
+
+* Registry entries specified on the command line.
+
+A registry is a JSON file that looks like this:
+
+```
+{
+    "version": 1,
+    "flakes": {
+        "nixpkgs": {
+            "uri": "github:NixOS/nixpkgs"
+        },
+        ...
+    }
+}
+```
+
+With this registry, flake references resolve as follows:
+
+* `nixpkgs` -> `github:NixOS/nixpkgs`
+* `nixpkgs/release-19.03` -> `github:NixOS/nixpkgs/release-19.03`
+* `nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c` -> `github:NixOS/nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c`
+
+The registries are searched in reverse order. Thus the local registry
+overrides the global registry, and the command line takes precedence
+over the local and global registries.
```

Comment:
probably the same as:
```
nix-repl> builtins.fromJSON ''{ "foo": 3, "foo": 4 }''
{ foo = 4; }
```

Reply:
Should definitely be documented!

Reply:
Then `"flakes"` should be an array of objects in JSON terms, because objects are unorderdered and some implementations like databases will reorder the entries.